File: MyLeetCode/Leetcode/Longest Absolute File Path.cpp
/* Use a hashtable to report the len of the previous level
 *
 */
class Solution {
public:
    int lengthLongestPath(string input) {
        stringstream ss(input);
        string s;
        int maxlen = 0;
        unordered_map<int, int> table;
        table[0] = 0;
        
        while (getline(ss, s, '\n')) {
            int pos = s.find_first_not_of('\t');
            string name = s.substr(pos);
            int level = s.size() - name.size();
            if (name.find('.') == string::npos) {
                table[level + 1] = table[level] + name.size() + 1;
            }
            else {
                maxlen = max(maxlen, table[level] + (int)name.size()); 
            }
        }
        return maxlen;
    }
};

File: MyLeetCode/python/Longest Absolute File Path.py
###
# Hash Table
# Time Complexity: O(n) 
# Space Complexity: O(n)
###
class Solution(object):
    def lengthLongestPath(self, input):
        """
        :type input: str
        :rtype: int
        """
        if not input:
            return 0
        maxlen = 0
        d = {0:0}
        for line in input.splitlines():
            name = line.lstrip('\t')
            level = len(line) - len(name)
            if '.' not in name:
                d[level+1] = d[level] + len(name) + 1
            else:
                maxlen = max(maxlen, d[level] + len(name))
        return maxlen
        
        